### PR TITLE
Simplification of stupid model code

### PR DESCRIPTION
**Pull request by dybber**
I'm working on simplifying the stupid model code, still on my TODO is to simplify stupid12 to stupid16.

Some notes for further simplication of the code i've been through so far:
 - perhaps we shouldn't even have and use agents_random() in stupid09 and just use random.shuffle(), when model.agents is converted to a list?
 - perhaps stupid08, we shouldn't use a global variable to hold the file handle, but use a model variable?
 - would be nice if we could avoid update_current_tile()
 - would be nice if `Agent.align()` was named `Agent.center_in_tile()` or similar
 - perhaps all agents should have a reference to their model, instead of the requirement to pass model into all setup/step functions and into all "action" functions

### CODE REVIEW

File: examples/stupid/stupid01-basic.py [hunk 1]
```diff
@@ -1,43 +1,57 @@
...
     def setup(self, model):
         self.size = 8
         self.color = (255, 0, 0)
+
+        # Align to center of tile
         self.align()
+
+        # Optimization, to allow quick lookup of which agents are on a tile
+        # can we hide this somehow?
         self.update_current_tile()
```
Comment (JensKanstrupLarsen) [inline examples/stupid/stupid01-basic.py:15]: `update_current_tile` should probably just be called by the model when calling `add_agent`.

File: examples/stupid/stupid08-write-data-to-file.py [hunk 2]
```diff
@@ -4,79 +4,101 @@
...
+        bug_min = min(bug_min, agent.grow_size)
+        bug_max = max(bug_max, agent.grow_size)
+        bug_sum += agent.grow_size
+    bug_mean = bug_sum / len(model.agents)
+
+    # Write min, average and max bug size to file
+    f.write(str(bug_min) + " " + str(bug_mean) + " " + str(bug_max) + "\n")
+    # Flush is necessary as long as we can't call f.close()
+    # when the user exits the program
+    f.flush()
```
Comment (JensKanstrupLarsen) [inline examples/stupid/stupid08-write-data-to-file.py:98]: This should not be necessary anymore, as we have the `Model.on_close` method where a call to `f.close` can be added.

**Comment by JensKanstrupLarsen** [comment]
Looks good to me.

- Using `random.shuffle()` is probably the better solution. I have also submitted a pull-request where it is removed, so it might as well be the only solution.
- I think we can avoid having the user make any call to `Agent.update_current_tile()` if we just make the model itself call it when it adds an agent though `Model.add_agent()`.
- I actually thought `Agent.align()` had been renamed in another pull request, but apparently not. I agree that `Agent.center_in_tile()` or similar is better.
- All agents actually have a reference to their model in `Agent.__model` (it is set when calling `Model.add_agent()`), but it might as well just be renamed to `Agent.model` so we can remove the `model`-parameter required for the button functions.

File: examples/stupid/stupid01-basic.py (hunk 1 above)
Comment (dybber) [reply examples/stupid/stupid01-basic.py:15]: yeah, let's do that

File: examples/stupid/stupid08-write-data-to-file.py (hunk 2 above)
Comment (dybber) [reply examples/stupid/stupid08-write-data-to-file.py:98]: oh yeah, will look at that next time I'm working on this PR (you are also welcome to add commits to this branch though)

File: examples/stupid/stupid01-basic.py (hunk 1 above)
Comment (JensKanstrupLarsen) [reply examples/stupid/stupid01-basic.py:15]: Turns out the model actually already does that, so this is probably just a leftover from an earlier version.

File: examples/stupid/stupid08-write-data-to-file.py [hunk 3]
```diff
@@ -4,80 +4,101 @@
...
+    bug_mean = bug_sum / len(model.agents)
+
+    # Write min, average and max bug size to file
+    f.write(str(bug_min) + " " + str(bug_mean) + " " + str(bug_max) + "\n")
+
+    # Update plots
+    model.update_plots()
+
+def close(model):
+    global f
```
Comment (dybber) [inline examples/stupid/stupid08-write-data-to-file.py:100]: Her er `global` ikke nødvendigt

File: examples/stupid/stupid08-write-data-to-file.py [hunk 4]
```diff
@@ -93,13 +93,13 @@ def step(model):
...
     f.write(str(bug_min) + " " + str(bug_mean) + " " + str(bug_max) + "\n")
-    # Flush is necessary as long as we can't call f.close()
-    # when the user exits the program
-    f.flush()
 
     # Update plots
     model.update_plots()
 
+def close(model):
+    global f
```
Comment (dybber) [inline examples/stupid/stupid08-write-data-to-file.py:100]: Her er `global` ikke nødvendigt

File: examples/stupid/stupid09-agent-order-randomized.py [hunk 5]
```diff
@@ -93,13 +93,13 @@ def step(model):
...
     f.write(str(bug_min) + " " + str(bug_mean) + " " + str(bug_max) + "\n")
-    # Flush is necessary as long as we can't call f.close()
-    # when the user exits the program
-    f.flush()
 
     # Update plots
     model.update_plots()
 
+def close(model):
+    global f
```
Comment (dybber) [inline examples/stupid/stupid09-agent-order-randomized.py:100]: Samme gælder de efterfølgende eksempler - `global` er ikke nødvendigt, når du ikke skal assigne til variablen.

File: examples/stupid/stupid08-write-data-to-file.py (hunk 4 above)
Comment (dybber) [reply examples/stupid/stupid08-write-data-to-file.py:100]: Jeg fjerner det selv.

**Comment by dybber** [comment]
From StupidModel 12:

"• The model stopping rule is changed: the model stops after 1000 time steps have
been executed or when the number of bugs reaches zero."

Should we keep an iteration-count variable in the model? Currently I'm adding it as a TODO in a comment in stupid12-mortality-reproduction.py

**Comment by dybber** [comment]
I've been through the rest of the stupid models, trying to simplify everything. There's still some TODOs left (marked in comments). In addition to those TODO's everywhere we open a file, and have the on_close handler, we need to make it conditional on the file actually being open:

```python
def close(model):
    if file_handle:
        file_handle.close()
```

I've only done that in the last model (stupid16), the same fix could be applied in the rest of the models.

**Comment by dybber** [comment]
Perhaps some utility functions for working with list of agents would be nice:
 - does this list contain an agent of the specified type?
 - get all agents of a specified type from a list
 - sort by an attribute

Also, it's unintuitive that agents_ordered also removes agents that doesn't have the property. Perhaps it's better either to keep them/add them in the end or to error, that is, first let the user get all agents of a specific type and then call a sort function

**Pull request merged**